README
SVN::Look - A caching wrapper around the svnlook command.

The svnlook command is the workhorse of Subversion hook scripts, being
used to gather all sorts of information about a repository, its
revisions, and its transactions. This module provides a simple object
oriented interface to a specific svnlook invocation, to make it easier
to hook writers to get and use the information they need.

Copyright (c) 2012 by CPqD (http://www.cpqd.com.br/)

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

Home page: http://search.cpan.org/dist/SVN-Look/


Changes
Revision history for SVN-Look. -*- text -*-

{{$NEXT}}

0.43      2022-05-21 11:30:57-03:00 America/Sao_Paulo

[Fixes]

- Skip failing tests that depend on previous ones. Also, show diagnostics when
  tests fail.

0.42      2020-10-08 22:35:29-03:00 America/Sao_Paulo

[Fixes]

- Regenerate Makefile.PL with ExtUtils::MakeMaker version 7.48 in order to fix a
  bug in previous versions of it.

- Fix a few problems detected by PerlCritic.

0.41      2014-06-13 23:50:53-03:00 America/Sao_Paulo

[Fixes]

- Restore support for Perl 5.8.

0.40      2014-06-12 15:04:39-03:00 America/Sao_Paulo

[Fixes]

- Fix security issue when invoking external commands.

- Support base64 encoded property values.

- Don't strip trailing spaces from filenames grokked by the changed_hash
  method.

[Thanks]

- Many thanks to Stephane Chazelas who reported the issues fixed in this
  version.

0.39      2013-10-20 21:31:50 America/Sao_Paulo

0.38      2012-06-19 14:00:00 America/Sao_Paulo

	Method "changed" now returns the entries sorted and with no
	duplicates.

0.37      2012-06-18 14:34:31 America/Sao_Paulo

	Revert previous proplist fix. It wasn't really a fix.

0.36      2012-06-18 14:05:44 America/Sao_Paulo

	Fix proplist, which died when invoked on an artifact with no
	properties. Now it returns an empty hash.

	Improve error messages upon svnlook command failure.

0.35      2012-04-21 21:13:18 America/Sao_Paulo

	Moved from Google Code to GitHub.

0.34      2012-03-05 11:35:50 America/Sao_Paulo

	Corrects RT#75545 and RT#75546 to make it supported in
	Windows. (Thanks Mike Brimer!)

0.33      2012-02-27 20:48:45 America/Sao_Paulo

	Require Perl 5.8.0.

0.32      2012-02-26 18:54:21 America/Sao_Paulo

	Reinsert check for svn command in Makefile.PL.

0.31      2012-02-25 22:02:07 America/Sao_Paulo

0.30	2011-11-02

	Change the call to svnlook to use execvp directly and avoiding
	the shell.

	Moved the repository to git (http://code.google.com/p/svn-look/source).

0.29    2011-10-12

	Implement method 'filesize', new with Subversion 1.7.0.

0.28    2011-10-10

	Reimplementing methods 'repo', 'rev', and 'txn'. They are used
	by SVN::Hooks::Notify, as I have been kindly reminded by Bill
	Moseley. Thanks!

0.27    2011-09-17

	Make the arguments -r/-t optional.
	Remove unecessary methods 'repo', 'rev', and 'txn'.
	Implement methods 'date', 'info', and 'propget'.

0.26    2011-08-27

	Requiring at least svnlook version 1.4.0.

0.25	2011-08-21

	Method ::cat don't chomp the file contents when called in
	scalar context anymore.

	Don't append hardcoded paths to $ENV{PATH} anymore.

0.24	2011-07-30

	Tests grok username using svnlook to be sure.

0.23	2011-07-27

	Porting Makefile.PL to Windows.
	Adding URI::file to the dependencies.

0.22	2011-07-21

	Avoiding dependency from File::Slurp in the tests.

0.21	2011-07-20

	Ported to Windows with Strawberry Perl and also with Active Perl.

0.20	2010-12-12

	Fix methods copied_from and copied_to.

0.19	2010-12-12

	Implements the method tree.

0.18	2010-04-28

	Being more careful during tests.

0.17    2010-02-24

	Minor error correction.

0.16	2010-02-16

	Implements the methods: youngest, uuid, and lock.

	Bypasses an issue in the testing of method author.

	Croaks instead of dying.

0.15	2009-10-24

	Drops the sub-minor version. Now it's just M.mm.

	Implements the kwalitee test.

	Implements the --author-tests flag to Makefile.PL, which
	enables the pod and kwalitee tests.

0.13	2008-11-06

	Changing the LICENSE from GPL to Perl.

0.12	2008-10-09

	Removing dependency from SVN::Hooks from the tests.

0.11	2008-10-04

	Using File::Spec to be a little more portable.

0.10	2008-09-27

	Implemented the 'diff' method.

0.09	2008-09-27
	Removed dependency from Switch.pm.

	Look for the svnlook command in a bunch of directories.

	Now the module can be built without the svn commands but the
	tests will check this and be skipped.

	Added tests: t/pod.t and t/pod-coverage.t

	Private functions got the _ prefix to avoid complaints from
	pod-coverage.t. 

0.08    2008-09-25/23:00:00
        First version released independently of SVN::Hooks.




t/test-functions.pl
# Copyright (C) 2008 by CPqD

use strict;
use warnings;
use Cwd;
use File::Temp qw/tempdir/;
use File::Spec::Functions;
use File::Path;
use File::Copy;
use URI::file;

# Make sure the svn messages come in English.
$ENV{LC_MESSAGES} = 'C';

sub can_svn {
  CMD:
    for my $cmd (qw/svn svnadmin svnlook/) {
	eval {
	    open my $pipe, '-|', "$cmd --version" or die;
	    local $/ = undef;		# slurp mode
	    <$pipe>;
	    close $pipe or die;
	};
	return 0 if $@;
    }
    return 1;
}

our $T;

sub newdir {
    my $num = 1 + Test::Builder->new()->current_test();
    my $dir = catdir($T, $num);
    mkdir $dir;
    $dir;
}

sub do_script {
    my ($dir, $cmd) = @_;
    my $script = catfile($dir, 'script');
    my $stdout = catfile($dir, 'stdout');
    my $stderr = catfile($dir, 'stderr');
    {
	open my $fd, '>', $script or die;
	print $fd $cmd;
	close $fd;
	chmod 0755, $script;
    }
    copy(catfile($T, 'repo', 'hooks', 'svn-hooks.pl')   => catfile($dir, 'svn-hooks.pl'));
    copy(catfile($T, 'repo', 'conf',  'svn-hooks.conf') => catfile($dir, 'svn-hooks.conf'));

    system("$script 1>$stdout 2>$stderr");
}

sub read_file {
    my ($file) = @_;
    open my $fd, '<', $file or die "Can't open '$file': $!\n";
    local $/ = undef;		# slurp mode
    return <$fd>;
}

sub work_ok {
    my ($tag, $cmd) = @_;
    my $dir = newdir();
    ok((do_script($dir, $cmd) == 0), $tag)
	or diag("work_ok command failed with following stderr:\n",
		scalar(read_file(catfile($dir, 'stderr'))));
}

sub work_nok {
    my ($tag, $error_expect, $cmd) = @_;
    my $dir = newdir();
    my $exit = do_script($dir, $cmd);
    if ($exit == 0) {
	fail($tag);
	diag("work_nok command worked but it shouldn't!\n");
	return;
    }

    my $stderr = scalar(read_file(catfile($dir, 'stderr')));

    if (! ref $error_expect) {
	ok(index($stderr, $error_expect) >= 0, $tag)
	    or diag("work_nok:\n  '$stderr'\n    does not contain\n  '$error_expect'\n");
    }
    elsif (ref $error_expect eq 'Regexp') {
	like($stderr, $error_expect, $tag);
    }
    else {
	fail($tag);
	diag("work_nok: invalid second argument to test.\n");
    }
}

sub reset_repo {
    my $cleanup = exists $ENV{REPO_CLEANUP} ? $ENV{REPO_CLEANUP} : 1;
    $T = tempdir('t.XXXX', DIR => getcwd(), CLEANUP => $cleanup);

    my $repo = catfile($T, 'repo');
    my $wc   = catfile($T, 'wc');

    system("svnadmin create $repo");

    my $repouri = URI::file->new($repo);

    system("svn co -q $repouri $wc");

    return $T;
}

1;


Makefile.PL
# This file was generated by http://search.cpan.org/dist/Dist-Zilla-Plugins-CJM/lib/Dist/Zilla/Plugin/MakeMaker/Custom.pm

use strict;
use warnings;

# Check if we have a suitable version of the svnlook command.
eval {
    my $path = $ENV{PATH} || '';
    open my $svnlook, '-|', 'svnlook --version'
        or die "Aborting because I couldn't find the 'svnlook' executable in PATH='$path'.\n";
    $_ = <$svnlook>;
    if (my ($major, $minor, $patch) = (/(\d+)\.(\d+)\.(\d+)/)) {
        $major > 1 || $major == 1 && $minor >= 4
            or die "I need at least version 1.4.0 of svnlook but you have only $major.$minor.$patch.\n";
    } else {
        die "Can't grok Subversion version from svnlook --version command.\n";
    }
    local $/ = undef;           # slurp mode
    <$svnlook>;
    close $svnlook or die "Can't close svnlook commnand.\n";
};
if ($@) {
    warn $@;
    # CPAN Testers says
    # (http://wiki.cpantesters.org/wiki/CPANAuthorNotes) that we
    # should simply exit normally before generating the Makefile if
    # there is an unmet dependency.
    exit 0;
}

use ExtUtils::MakeMaker ##{ $eumm_version ##};

my %WriteMakefileArgs = (
    ##{ $plugin->get_default(keys %default_args) ##}
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/SVN/Look.pm
## no critic (Modules::RequireExplicitPackage, InputOutput::RequireBriefOpen)

use 5.008;
use strict;
use warnings;

package SVN::Look;
# ABSTRACT: Caching wrapper around the svnlook command

use Carp;
use File::Spec::Functions;
use List::MoreUtils qw{uniq};

=head1 SYNOPSIS

  use SVN::Look;
  my $revlook = SVN::Look->new('/repo/path', -r => 123);
  my $author  = $revlook->author();
  my $msg     = $revlook->log_msg();
  my @added_files   = $revlook->added();
  my @updated_files = $revlook->updated();
  my @deleted_files = $revlook->deleted();
  my @changed_files = $revlook->changed();
  my $file_contents = $revlook->cat('/path/to/file/in/repository');

  my $trxlook = SVN::Look->new('/repo/path', -t => 123);

=head1 DESCRIPTION

The svnlook command is the workhorse of Subversion hook scripts, being
used to gather all sorts of information about a repository, its
revisions, and its transactions. This module provides a simple object
oriented interface to a specific svnlook invocation, to make it easier
to hook writers to get and use the information they need. Moreover,
all the information gathered by calling the svnlook command is cached
in the object, avoiding repetitious calls.

=cut

my @SVN_VERSION;

BEGIN {
    my $path = $ENV{PATH} || '';

    # Perl on Windows doesn't support the piped three-arg open. So we use
    # the two-arg open here and don't care about Perl::Critic since this is
    # a fixed string command.

    open my $svnlook, 'svnlook --version |' ## no critic (InputOutput::ProhibitTwoArgOpen)
        or die "Aborting because I couldn't find the 'svnlook' executable in PATH='$path'.\n";
    $_ = <$svnlook>;
    if (@SVN_VERSION = (/(\d+)\.(\d+)\.(\d+)/)) {
        if ($SVN_VERSION[0] < 1 || $SVN_VERSION[0] == 1 && $SVN_VERSION[1] < 4) {
            die "I need at least version 1.4.0 of svnlook but you have only ",
                join('.', @SVN_VERSION), " in PATH='$path'.\n";
        }
    } else {
        die "Can't grok Subversion version from svnlook --version command.\n";
    }
    local $/ = undef;           # slurp mode
    <$svnlook>;
    close $svnlook or die "Can't close svnlook command.\n";
}

=method B<new> REPO [, WHAT, NUMBER]

The SVN::Look constructor needs one or three arguments:

=over

=item REPO is the path to the repository.

=item WHAT must be either '-r' or '-t', specifying if the third
argument is a revision number or a transaction number, respectively.
If neither -r or -t is specified, the HEAD revision is used.

=item NUMBER is either a revision or transaction NUMBER, as specified
by WHAT.

=back

=cut

sub new {
    my ($class, $repo, @opts) = @_;
    my $self = {
        repo => $repo,
        opts => [@opts],
    };
    bless $self, $class;
    return $self;
}

sub _svnlook {
    my ($self, $cmd, @args) = @_;
    my @cmd = ('svnlook', $cmd, $self->{repo});
    push @cmd, @{$self->{opts}} unless $cmd =~ /^(?:youngest|uuid|lock)$/;

    my $fd;
    my $tmpfile;
    if ($^O ne 'MSWin32') {
        open $fd, '-|', @cmd, @args
            or die "Can't exec svnlook $cmd: $!\n";
    } else {
        # Windows doesn't support the three-argument version of open
        # neither the pipe function. So we run the svnlook command
        # with system, sending its output to a temporary file and
        # opening the file later in $fd.

        # Create the temporary file.
        require File::Temp;
        $tmpfile = File::Temp->new();
        my $filename = $tmpfile->filename;

        ## no critic (ProhibitTwoArgOpen, ProhibitBarewordFileHandles)

        # Dup STDOUT and direct it to the file
        no warnings 'once';     ## no critic (TestingAndDebugging::ProhibitNoWarnings)
        open OLDOUT, '>&STDOUT'   or die "Can't dup STDOUT: $!\n";
        open STDOUT, ">$filename" or die "Can't redirect STDOUT to $filename: $!\n";

        # Shell out the svnlook command
        system(@cmd, @args) == 0
            or die "system @cmd failed: $?\n";

        # Restore STDOUT
        open STDOUT, '>&OLDOUT' or die "Can't redirect STDOUT to its former value: $!\n";

        # Open the temporary file
        open $fd, "<$filename" or die "Can't open $filename: $!\n";

        ## use critic
    }

    if (wantarray) {
        my @lines = <$fd>;
        unless (close $fd) {
            if ($!) {
                die "Error closing (wantarray) svnlook $cmd pipe: $!\n";
            } else {
                die "Exit status $? from (wantarray) svnlook $cmd\n";
            }
        }
        chomp @lines;
        return @lines;
    }
    else {
        local $/ = undef;
        my $line = <$fd>;
        unless (close $fd) {
            if ($!) {
                die "Error closing svnlook $cmd pipe: $!\n";
            } else {
                die "Exit status $? from svnlook $cmd\n";
            }
        }
        chomp $line unless $cmd eq 'cat';
        return $line;
    }
}

=method B<repo>

Returns the repository path that was passed to the constructor.

=cut

sub repo {
    my $self = shift;
    return $self->{repo};
}

=method B<txn>

Returns the transaction number that was passed to the constructor. If
none was passed, returns undef.

=cut

sub txn {
    my $self = shift;
    return $self->{opts}[0] eq '-t' ? $self->{opts}[1] : undef;
}

=method B<rev>

Returns the revision number that was passed to the constructor. If
none was passed, returns undef.

=cut

sub rev {
    my $self = shift;
    return $self->{opts}[0] eq '-r' ? $self->{opts}[1] : undef;
}

=method B<author>

Returns the author of the revision/transaction.

=cut

sub author {
    my $self = shift;
    unless (exists $self->{author}) {
        chomp($self->{author} = $self->_svnlook('author'));
    }
    return $self->{author};
}

=method B<cat> PATH

Returns the contents of the file at PATH. In scalar context, return
the whole contents in a single string. In list context returns a list
of chomped lines.

=cut

sub cat {
    my ($self, $path) = @_;
    return $self->_svnlook('cat', $path);
}

=method B<changed_hash>

Returns a reference to a hash containing information about all file
changes occurred in the revision. The hash always has the following
keys:

=over

=item added

A list of files added in the revision.

=item deleted

A list of files deleted in the revision.

=item updated

A list of files updated in the revision.

=item prop_modified

A list of files that had properties modified in the revision.

=item copied

A hash containing information about each file or diretory copied in the revision. The hash keys are the names of elements copied to. The value associated with a key is a two-element array containing the name of the element copied from and the specific revision from which it was copied.

=back

=cut

sub changed_hash {
    my $self = shift;
    unless ($self->{changed_hash}) {
        my (@added, @deleted, @updated, @prop_modified, %copied);
        foreach ($self->_svnlook('changed', '--copy-info')) {
            next if length($_) <= 4;
            chomp;
            my ($action, $prop, $changed) = unpack 'AAxx a*', $_;
            if    ($action eq 'A') {
                push @added,   $changed;
            }
            elsif ($action eq 'D') {
                push @deleted, $changed;
            }
            elsif ($action eq 'U') {
                push @updated, $changed;
            }
            else {
                if ($changed =~ /^\(from (.*?):r(\d+)\)$/) {
                    $copied{$added[-1]} = [$1 => $2];
                }
            }
            if ($prop eq 'U') {
                push @prop_modified, $changed;
            }
        }
        $self->{changed_hash} = {
            added         => \@added,
            deleted       => \@deleted,
            updated       => \@updated,
            prop_modified => \@prop_modified,
            copied        => \%copied,
        };
    }
    return $self->{changed_hash};
}

=method B<added>

Returns the list of files added in the revision/transaction.

=cut

sub added {
    my $self = shift;
    return @{$self->changed_hash()->{added}};
}

=method B<updated>

Returns the list of files updated in the revision/transaction.

=cut

sub updated {
    my $self = shift;
    return @{$self->changed_hash()->{updated}};
}

=method B<deleted>

Returns the list of files deleted in the revision/transaction.

=cut

sub deleted {
    my $self = shift;
    return @{$self->changed_hash()->{deleted}};
}

=method B<prop_modified>

Returns the list of files that had properties modified in the
revision/transaction.

=cut

sub prop_modified {
    my $self = shift;
    return @{$self->changed_hash()->{prop_modified}};
}

=method B<changed>

Returns the list of all files added, updated, deleted, and the ones
that had properties modified in the revision/transaction.

=cut

sub changed {
    my $self = shift;
    my $hash = $self->changed_hash();
    unless (exists $hash->{changed}) {
        $hash->{changed} = [sort(uniq(@{$hash->{added}},
                                      @{$hash->{updated}},
                                      @{$hash->{deleted}},
                                      @{$hash->{prop_modified}}))];
    }
    return @{$hash->{changed}};
}

=method B<copied_to>

Returns the list of new names of files that were copied in the
revision/transaction.

=cut

sub copied_to {
    my $self = shift;
    return keys %{$self->changed_hash()->{copied}};
}

=method B<copied_from>

Returns the list of original names of files that were copied in the
revision/transaction. The order of this list is guaranteed to agree
with the order generated by the method copied_to.

=cut

sub copied_from {
    my $self = shift;
    return map {$_->[0]} values %{$self->changed_hash()->{copied}};
}

=method B<date>

Returns the date of the revision/transaction.

=cut

sub date {
    my $self = shift;
    unless (exists $self->{date}) {
        $self->{date} = $self->_svnlook('date');
    }
    return $self->{date};
}

=method B<diff> [OPTS, ...]

Returns the GNU-style diffs of changed files and properties. There are
three optional options that can be passed as strings:

=over

=item C<--no-diff-deleted>

Do not print differences for deleted files

=item C<--no-diff-added>

Do not print differences for added files.

=item C<--diff-copy-from>

Print differences against the copy source.

=back

In scalar context, return the whole diff in a single string. In list
context returns a list of chomped lines.

=cut

sub diff {
    my ($self, @opts) = @_;
    return $self->_svnlook('diff', @opts);
}

=method B<dirs_changed>

Returns the list of directories changed in the revision/transaction.

=cut

sub dirs_changed {
    my $self = shift;
    unless (exists $self->{dirs_changed}) {
        my @dirs = $self->_svnlook('dirs-changed');
        $self->{dirs_changed} = \@dirs;
    }
    return @{$self->{dirs_changed}};
}

=method B<filesize> PATH

Returns the size (in bytes) of the file located at PATH as it is
represented in the repository.

=cut

sub filesize {
    my ($self, $path) = @_;
    return $self->_svnlook('filesize', $path);
}

=method B<info>

Returns the author, datestamp, log message size, and log message of
the revision/transaction.

=cut

sub info {
    my $self = shift;
    return $self->_svnlook('info');
}

=method B<lock> PATH

If PATH has a lock, returns a hash containing information about the lock, with the following keys:

=over

=item UUID Token

A string with the opaque lock token.

=item Owner

The name of the user that has the lock.

=item Created

The time at which the lock was created, in a format like this: '2010-02-16 17:23:08 -0200 (Tue, 16 Feb 2010)'.

=item Comment

The lock comment.

=back

If PATH has no lock, returns undef.

=cut

sub lock {                      ## no critic (ProhibitBuiltinHomonyms)
    my ($self, $path) = @_;
    my %lock = ();
    my @lock = $self->_svnlook('lock', $path);

    while (my $line = shift @lock) {
        chomp $line;
        my ($key, $value) = split /:\s*/, $line, 2;
        if ($key =~ /^Comment/) {
            $lock{Comment} = join('', @lock);
            last;
        }
        else {
            $lock{$key} = $value;
        }
    }

    return %lock ? \%lock : undef;
}

=method B<log_msg>

Returns the log message of the revision/transaction.

=cut

sub log_msg {
    my $self = shift;
    unless (exists $self->{log}) {
        $self->{log} = $self->_svnlook('log');
    }
    return $self->{log};
}

=method B<propget> PROPNAME PATH

Returns the value of PROPNAME in PATH.

=cut

sub propget {
    my ($self, @args) = @_;
    return $self->_svnlook('propget', @args);
}

=method B<proplist> PATH

Returns a reference to a hash containing the properties associated with PATH.

=cut

# The 'svnlook proplist' command had its output format changed in svn
# 1.8.0. So, in order to make the code more stable we try to use the
# --xml option. However, this option was implemented on svn 1.6.0
# only. Since we still want to support older Subversions we have to
# check if we can use that option first. See discussion on
# https://github.com/gnustavo/SVN-Look/pull/1.

my $proplist_does_support_xml_option;

sub proplist {
    my ($self, $path) = @_;

    unless ($self->{proplist}{$path}) {
        $proplist_does_support_xml_option = $SVN_VERSION[0] > 1 || $SVN_VERSION[0] == 1 && $SVN_VERSION[1] >= 8
            unless defined $proplist_does_support_xml_option;

        if ($proplist_does_support_xml_option) {
            my $xml = $self->_svnlook(qw/proplist --verbose --xml/, $path);
            require XML::Simple;
            my $dom = XML::Simple::XMLin($xml, ForceArray => ['property']);
            while (my ($prop, $value) = each %{$dom->{target}{property}}) {
                my $content = $value->{content};
                if (my $encoding = $value->{encoding}) {
                    if ($encoding eq 'base64') {
                        require MIME::Base64;
                        $content = MIME::Base64::decode($content);
                    } else {
                        die "Don't know how to decode property '$prop' value encoded as '$encoding'\n";
                    }
                }
                $self->{proplist}{$path}{$prop} = $content;
            }
        } else {
            # Old syntax up to SVN 1.7.
            my $text = $self->_svnlook('proplist', '--verbose', $path);
            my @list = split /^\s\s(\S+)\s:\s/m, $text;
            shift @list;        # skip the leading empty field
            chomp(my %hash = @list);
            $self->{proplist}{$path} = \%hash;
        }
    }
    return $self->{proplist}{$path};
}

=method B<tree> [PATH_IN_REPOS, OPTS, ...]

Returns the repository tree as a list of paths, starting at
PATH_IN_REPOS (if supplied, at the root of the tree otherwise),
optionally showing node revision ids.

=over

=item C<--full-paths>

show full paths instead of indenting them.

=item C<--show-ids>

Returns the node revision ids for each path.

=item C<--non-recursive>

Operate on single directory only.

=back

=cut

sub tree {
    my ($self, @args) = @_;
    return $self->_svnlook('tree', @args);
}

=method B<uuid>

Returns the repository's UUID.

=cut

sub uuid {
    my ($self) = @_;
    return $self->_svnlook('uuid');
}

=method B<youngest>

Returns the repository's youngest revision number.

=cut

sub youngest {
    my ($self) = @_;
    return $self->_svnlook('youngest');
}

1; # End of SVN::Look

=head1 REPOSITORY

L<https://github.com/gnustavo/SVN-Look>
